lib/OpenQA/Schema/Result/Jobs.pm

```diff
@@ -229,26 +229,39 @@ sub sqlt_deploy_hook {
     $sqlt_table->add_index(name => 'idx_jobs_scenario',    fields => [qw(VERSION DISTRI FLAVOR TEST MACHINE ARCH)]);
 }
 
-# override to straighten out job modules
+# override to straighten out job modules and screenshots
 sub delete {
     my ($self) = @_;
 
+    my $schema = $self->result_source->schema;
+    my $dbh    = $schema->storage->dbh;
+
+    # delete all job modules
     $self->modules->delete;
 
-    my $sls = $self->screenshot_links->search({}, {columns => 'screenshot_id'});
-    my %ids = map { $_->screenshot_id => 1 } $sls->all;
-    # delete all references
-    $self->screenshot_links->delete;
-    my $schema = $self->result_source->schema;
+    # unlink screenshots
+    my $screenshots
+      = $schema->resultset('Screenshots')->search({job_id => $self->id}, {join => 'links'});
+    my $decrement_link_counter = $dbh->prepare('UPDATE screenshots SET link_count = link_count - 1 WHERE id = ?');
+    while (my $screenshot = $screenshots->next) {
+        my $link_count = $screenshot->link_count;
+
+        # stop here while migration for link_count still ongoing
+        return 0 unless defined $link_count;
 
-    $sls = $schema->resultset('ScreenshotLinks')
-      ->search({screenshot_id => {-in => [sort keys %ids]}}, {columns => 'screenshot_id', distinct => 1});
-    map { delete $ids{$_->screenshot_id} } $sls->all;
+        # delete exclusively used screenshots
+        if ($link_count <= 1) {
```

Comment:
this doesn't look any more performant than what we have, sorry :(

If you want performance you need to leave the complete job to the database - and that means requiring the migration to be run. I'd rather take a one maintenance window with a long running migration than a continously slow cleanup.

E.g. please benchmark `UPDATE screenshots SET link_count = link_count - 1 WHERE id in (select screenshot_id FROM screenshot_links WHERE job_id = ?`

Deleting the unreferenced screenshots can happen any time, you can actually make that a complete different task running on different schedule, they are clearly marked as having `linked_count=0`

Reply:
Ok, so you want the screenshot deletion happening in a separate (Minion) job and only decrement the link count here. That would be faster, indeed.

> this doesn't look any more performant than what we have, sorry :(

We don't get rid of the file system operations but the database query is now actually quite fast. It was 11 seconds on my system before alone for finding the exclusively used screenshots. Now it is just about 0.0001 seconds per screenshot in that loop.

Reply:
Then avoid the loop as well